Swiss public transport config flow

miaucl opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
Extend the swiss_public_transport integration with a config_flow and unit tests.

This is in preparation for https://github.com/home-assistant/core/pull/105264#issuecomment-1846225721


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/30340
- This PR is related to issue: https://github.com/home-assistant/core/pull/105264#issuecomment-1846225721

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

miaucl committed:
add config flow

miaucl committed:
unit tests

joostlek commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -6,45 +6,60 @@
+    )
 
 
 async def async_setup_platform(
```
So the thing is, we should not use YAML anymore. Instead we want to import the YAML using an import flow which creates a config entry for the YAML config.

miaucl replied:
@joostlek Do you mean, it should not be possible anymore to configure via `configuration.yaml`?

miaucl replied:
Ohh, that is sad...
But removing it will be a breaking change and I would rather not do that.

But I don't think this is up to me to decide. What you would say @fabaff ?

joostlek replied:
Well, we have an import flow for that. You can then import the YAML so it will create an entry for you

miaucl replied:
Ok, have not yet looked at that.

miaucl replied:
@joostlek Did you refer to this when you were talking about `import flow`?
https://github.com/home-assistant/example-custom-config/blob/master/custom_components/example_load_platform/README.md

joostlek replied:
Nope: checkout https://github.com/home-assistant/core/pull/104730/files#diff-3f88feab5a41e23e0057f3117af8a8effc16d9a136c2dbb07e293286a9acb0ab

(This is just to show how an import would work. I changed this code up a bit as its now a bit cleaner, so please checkout the version of suez water on dev)

miaucl replied:
Thanks for this hint. I have a new proposition for it based on the Suez example.

joostlek commented on homeassistant/components/swiss_public_transport/__init__.py:
```diff
@@ -1 +1,38 @@
+) -> bool:
+    """Set up platform from a ConfigEntry."""
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = entry.data
```
Can we initialize the api client here, do a quick check if the service is working and then store it in the data?

miaucl commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -0,0 +1,37 @@
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
```
You mean, using the `from` and `to` config values for example, to build a entry id?

miaucl commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -0,0 +1,37 @@
+    ) -> FlowResult:
+        """Async user step to set up the connection."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            return self.async_create_entry(title=user_input[CONF_NAME], data=user_input)
```
The api will match it to the closest station, even when I type `X` and `Y` for `from` and `to`, the api request will succeed.

miaucl replied:
Test added

miaucl committed:
yaml config import flow

joostlek commented on homeassistant/components/swiss_public_transport/__init__.py:
```diff
@@ -1,38 +1,75 @@
     return True
+
+
+async def swiss_public_transport_opendata_client(
```
You can just put this function in async_setup_entry, it's not that big (and it gives a better overview of the setup)

miaucl commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -30,8 +33,85 @@ async def async_step_user(
             step_id="user", data_schema=DATA_SCHEMA, errors=errors
         )
+
+    async def async_step_import(self, import_input: dict[str, Any]) -> FlowResult:
```
The import issues now use the same implementation as suez_water.

miaucl replied:
Hey, no stress. Already thank you for your in-depth help here. I am not quite sure how to type this or what to do differently compared to the suez_water example.

joostlek replied:
https://github.com/home-assistant/core/pull/105558

miaucl replied:
Thank you, I think I understood now what you meant. I adapted the code accordingly to the PR.

miaucl committed:
change deprecation period and simply code

joostlek commented on homeassistant/components/swiss_public_transport/__init__.py:
```diff
@@ -1 +1,64 @@
+    """Set up platform from a ConfigEntry."""
+    config = entry.data
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = config
```
Why do we store the config entry data here? This is also being passed on to the async setup entries in the other platforms, so is there any need to store it here?

joostlek commented on homeassistant/components/swiss_public_transport/__init__.py:
```diff
@@ -1 +1,64 @@
+        await opendata.async_get_data()
+    except OpendataTransportError:
+        _LOGGER.error(
+            "Check at http://transport.opendata.ch/examples/stationboard.html "
```
I think there are 2 things we should try and catch. Because we already check if the stations exist in the config flow, we know that _in theory_ the stations exist and that this can work. But there are 2 possibilities where we can't get data.

1. The service is down. In this case we want to raise a ConfigEntryNotReady exception. This makes it retry a bit later again.
2. The stations don't exist anymore. (I'm just doing a guess this is something that can happen). In this case, our config entry just doesn't work. In the future, if you feel really fancy you could work on raising a repair issue to notify the user that they should repair their config entry. For now, please log that the call failed and raise ConfigEntryError instead.

This is the ideal solution, but this also requires the library to raise different issues on the different causes. Is there any way you can maybe tinker around with this and check if we can catch specifics?

miaucl replied:
Yes, I see the point. There are two errors thrown by the library, one for connection problems: `OpendataTransportConnectionError`; and one, when he cannot interpret the response `OpendataTransportError`.

It can be found here:
https://github.com/home-assistant-ecosystem/python-opendata-transport/blob/2077fe0bfe4e4e71554d6d364c2b4a30c9729863/opendata_transport/__init__.py#L290

But I also managed to get a `KeyError` out of the library, which the library did not catch, and I would therefore propose to also add a catch all at the end for _other_ errors.

joostlek replied:
Beware, we may not do `except Exception:`, that's only allowed in the config flow.

miaucl replied:
Ok, so I will only take care of the Errors the library provides.

joostlek replied:
But, you can catch a KeyError (and I hope you will update the lib to catch this error properly in the lib and patch it up here)

miaucl replied:
I realized, I only got the KeyError because I used the `AiohttpClientMocker` in the unit test and provided invalid data. I would rather not implement the KeyError and patch the library to also catch KeyErrors.

joostlek commented on homeassistant/components/swiss_public_transport/__init__.py:
```diff
@@ -1 +1,64 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
Can be removed

joostlek commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -0,0 +1,117 @@
+        """Async user step to set up the connection."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
```
Please only keep stuff in the try block that can raise

joostlek commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -0,0 +1,117 @@
+                errors["base"] = "unknown"
+            else:
+                return self.async_create_entry(
+                    title=f"{DOMAIN}_{user_input[CONF_START]}_{user_input[CONF_DESTINATION]}",
```
I think only the start and destination separated by a space would be enough.

joostlek commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -4,74 +4,69 @@
 
-SCAN_INTERVAL = timedelta(seconds=90)
-
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
I think the platform schema should stay in this file

miaucl committed:
keep name for legacy yaml
- removing the name now would break current implementations
- it will be removed together with the deprectation of  yaml config flow

miaucl committed:
improve error handling, simpler unique_id, cleanup

miaucl committed:
simplify issues for yaml import flow

joostlek commented on homeassistant/components/swiss_public_transport/__init__.py:
```diff
@@ -1 +1,65 @@
+            f"Setup failed for entry '{start} {destination}' with invalid data"
+        ) from e
+
+    hass.data[DOMAIN][f"{entry.entry_id}_opendata_client"] = opendata
```
```suggestion
    hass.data[DOMAIN][entry.entry_id] = opendata
```

joostlek commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: Optional[dict[str, Any]] = None
```
```suggestion
        self, user_input: dict[str, Any] | None = None
```

joostlek commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+        await self.async_set_unique_id(import_input[CONF_NAME])
+        self._abort_if_unique_id_configured()
+
+        try:
```
Only put stuff in the try block that can raise

joostlek commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+        except OpendataTransportError:
+            return self.async_abort(reason="client")
+        except Exception:  # pylint: disable=broad-except
+            return self.async_abort(reason="unknown")
```
Please add a logger.exception

joostlek commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+        except Exception:  # pylint: disable=broad-except
+            return self.async_abort(reason="unknown")
+
+        async_create_issue(
+            self.hass,
+            HOMEASSISTANT_DOMAIN,
+            f"deprecated_yaml_{DOMAIN}",
+            breaks_in_ha_version="2024.7.0",
+            is_fixable=False,
+            issue_domain=DOMAIN,
+            severity=IssueSeverity.WARNING,
+            translation_key="deprecated_yaml",
+            translation_placeholders={
+                "domain": DOMAIN,
+                "integration_title": DOMAIN,
+            },
+        )
```
You already have this one iirc

joostlek commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -4,74 +4,110 @@
     {
-        vol.Required(CONF_DESTINATION): cv.string,
+        vol.Required(CONF_NAME, default=DEFAULT_NAME): cv.string,
         vol.Required(CONF_START): cv.string,
-        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Required(CONF_DESTINATION): cv.string,
```
Please check the old code, it's slightly changed now

joostlek commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -4,74 +4,110 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the sensor from a config entry created in the integrations UI."""
+    opendata = hass.data[DOMAIN][f"{config_entry.entry_id}_opendata_client"]
```
```suggestion
    opendata = hass.data[DOMAIN][config_entry.entry_id]
```

joostlek commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -4,74 +4,110 @@
+    opendata = hass.data[DOMAIN][f"{config_entry.entry_id}_opendata_client"]
+
+    name = config_entry.data.get(CONF_NAME)
+    start = config_entry.data.get(CONF_START)
+    destination = config_entry.data.get(CONF_DESTINATION)
```
```suggestion
    start = config_entry.data[CONF_START]
    destination = config_entry.data[CONF_DESTINATION]
```

joostlek commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -4,74 +4,110 @@
+    """Set up the sensor from a config entry created in the integrations UI."""
+    opendata = hass.data[DOMAIN][f"{config_entry.entry_id}_opendata_client"]
+
+    name = config_entry.data.get(CONF_NAME)
```
```suggestion
    ---- put the start and destination here
    name = config_entry.data.get(CONF_NAME, f"{start} {destination}")
```

joostlek commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -86,12 +122,12 @@ def __init__(self, opendata, start, destination, name):
         self._from = start
         self._to = destination
-        self._remaining_time = ""
+        self._remaining_time = None
```
Why did this change btw

miaucl replied:
Because should be typed an `int` and `""` is misused as `None` here.

Here it is checked if it is set and smaller 0.
https://github.com/miaucl/ha/blob/994296031f00c785d1a7999f713d5eac5ee42952/homeassistant/components/swiss_public_transport/sensor.py#L168

miaucl replied:
This edge case was not present with the old yaml config, but with the config_flow, it seems to have a different order in calling the functions of the sensor, which is why I had to adapt it.

gjohansson-ST replied:
Should be `time` and not `int` right?
Would be good to actually type and do it also in the library properly

joostlek reviewed (changes requested):
You're making good progress :)

miaucl committed:
improve typing and clean name handling

joostlek commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -0,0 +1,82 @@
+            try:
+                await opendata.async_get_data()
+            except OpendataTransportError:
+                errors["base"] = "client"
```
I would prefer to be named "cannot_connect" to keep in line with other integrations as "client" doesn't really say what's up

joostlek commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -4,74 +4,110 @@
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
     {
-        vol.Required(CONF_DESTINATION): cv.string,
         vol.Required(CONF_START): cv.string,
```
```suggestion
        vol.Required(CONF_DESTINATION): cv.string,
```

joostlek commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -4,74 +4,110 @@
-        vol.Required(CONF_DESTINATION): cv.string,
         vol.Required(CONF_START): cv.string,
-        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Required(CONF_DESTINATION): cv.string,
+        vol.Required(CONF_NAME, default=DEFAULT_NAME): cv.string,
```
```suggestion
        vol.Required(CONF_START): cv.string,
        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
```

joostlek commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -4,74 +4,110 @@
+            translation_key="deprecated_yaml",
+            translation_placeholders={
+                "domain": DOMAIN,
+                "integration_title": DOMAIN,
```
The domain is usually lowercase and stuff, can we use the name used for humans?

joostlek commented on homeassistant/components/swiss_public_transport/strings.json:
```diff
@@ -0,0 +1,27 @@
+{
+  "config": {
+    "error": {
+      "client": "Invalid request: Check at [stationboard](http://transport.opendata.ch/examples/stationboard.html) if your station names are valid",
+      "unknown": "Could not parse request response: Check at [stationboard](http://transport.opendata.ch/examples/stationboard.html) if your station names are valid"
```
Both are kinda saying the same thing. I can imagine when you fill in non existent data it will only use of the two?

joostlek commented on homeassistant/components/swiss_public_transport/strings.json:
```diff
@@ -0,0 +1,27 @@
+    "step": {
+      "user": {
+        "data": {
+          "from": "Start Station",
```
Please use sentence case for translations

```suggestion
          "from": "Start station",
```

joostlek commented on homeassistant/components/swiss_public_transport/strings.json:
```diff
@@ -0,0 +1,27 @@
+    }
+  },
+  "issues": {
+    "deprecated_yaml_import_issue_client": {
```
Aren't you missing unknown?

joostlek commented on tests/components/swiss_public_transport/test_config_flow.py:
```diff
@@ -0,0 +1,228 @@
+
+async def test_flow_user_init(hass: HomeAssistant, aioclient_mock: AiohttpClientMocker):
+    """Test the initialization of the form in the first step of the config flow."""
+    aioclient_mock.get(
+        "http://transport.opendata.ch/v1/connections?from=test_start&to=test_destination",
+        json={"from": "test_start", "to": "test_destination", "connections": []},
+    )
+    result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN, context={"source": "user"}
+    )
+
```
Can we instead of mocking the http call mock the library? This way we're only testing the HA code

joostlek commented on tests/components/swiss_public_transport/test_config_flow.py:
```diff
@@ -0,0 +1,228 @@
+pytestmark = pytest.mark.usefixtures("mock_setup_entry")
+
+
+async def test_flow_user_init(hass: HomeAssistant, aioclient_mock: AiohttpClientMocker):
```
You can merge this test with the next one imo

miaucl replied:
I would rather prefer test one branch per unit test, no?

joostlek replied:
I usually just create 1 test for the full config flow (the happy path), the errors in the flow and then the same for the import flow

joostlek commented on tests/components/swiss_public_transport/test_config_flow.py:
```diff
@@ -0,0 +1,228 @@
+    )
+
+    assert result["type"] == "form"
+    assert result["errors"]["base"] == "client"
```
Can we also test the happy flow after an error? This way we also test if the config flow is able to recover from an error

joostlek commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -0,0 +1,82 @@
+
+    async def async_step_import(self, import_input: dict[str, Any]) -> FlowResult:
+        """Async import step to set up the connection."""
+        await self.async_set_unique_id(import_input[CONF_NAME])
+        self._abort_if_unique_id_configured()
+
```
Can we instead compare the start and destination data? Because we don't set unique id in the normal flow, it will lead to some inconsistencies.

Out the top of my head there's a self._async_abort_entries_match, do a search and find how it's used

joostlek reviewed (changes requested):
We're getting at the small details now :))

miaucl committed:
streamline unit tests
- happy path + errors
- mock opendata instead of aiohttp

joostlek commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -4,74 +4,111 @@
         vol.Required(CONF_DESTINATION): cv.string,
         vol.Required(CONF_START): cv.string,
-        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Required(CONF_NAME, default=DEFAULT_NAME): cv.string,
```
```suggestion
        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
```

joostlek commented on tests/components/swiss_public_transport/test_config_flow.py:
```diff
@@ -0,0 +1,295 @@
+    ) as mock_OpendataTransport:
+        mock_OpendataTransport_instance = mock_OpendataTransport.return_value
+        mock_OpendataTransport_instance.async_get_data.return_value = True
+        _result = await hass.config_entries.flow.async_init(
+            config_flow.DOMAIN, context={"source": "user"}
+        )
+        result = await hass.config_entries.flow.async_configure(
+            _result["flow_id"],
```
```suggestion
        result = await hass.config_entries.flow.async_init(
            config_flow.DOMAIN, context={"source": "user"}
        )
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
```

joostlek commented on tests/components/swiss_public_transport/test_config_flow.py:
```diff
@@ -0,0 +1,295 @@
+        assert result["type"] == "create_entry"
+        assert result["result"].title == "test_start test_destination"
+
+        assert {
+            CONF_START: "test_start",
+            CONF_DESTINATION: "test_destination",
+        } == result["data"]
```
Please put the data your testing at the left side and your expectation at the right side

joostlek commented on tests/components/swiss_public_transport/test_config_flow.py:
```diff
@@ -0,0 +1,295 @@
+        mock_OpendataTransport_instance = mock_OpendataTransport.return_value
+        mock_OpendataTransport_instance.async_get_data.side_effect = (
+            OpendataTransportError()
+        )
+        _result = await hass.config_entries.flow.async_init(
+            config_flow.DOMAIN, context={"source": "user"}
+        )
+        result = await hass.config_entries.flow.async_configure(
+            _result["flow_id"],
+            user_input={
+                CONF_START: "test_start",
+                CONF_DESTINATION: "test_destination",
+            },
+        )
+
+        assert result["type"] == "form"
+        assert result["errors"]["base"] == "bad_config"
+
+
+async def test_flow_user_init_data_unknown_error(hass: HomeAssistant) -> None:
+    """Test unknown errors."""
+    with patch(
+        "homeassistant.components.swiss_public_transport.config_flow.OpendataTransport",
+        autospec=True,
+    ) as mock_OpendataTransport:
+        mock_OpendataTransport_instance = mock_OpendataTransport.return_value
+        mock_OpendataTransport_instance.async_get_data.side_effect = IndexError()
+        _result = await hass.config_entries.flow.async_init(
+            config_flow.DOMAIN, context={"source": "user"}
+        )
+        result = await hass.config_entries.flow.async_configure(
+            _result["flow_id"],
+            user_input={
+                CONF_START: "test_start",
+                CONF_DESTINATION: "test_destination",
+            },
+        )
+
+        assert result["type"] == "form"
+        assert result["errors"]["base"] == "unknown"
```
So there's this wonderful thing called `@pytest.mark.parametrize` it allows you to parametrize your tests. You can use this to make the raised exception and the error message a parameter, which results in shorter code :)

miaucl replied:
Uhh, nice, that is fancy. Will try it :)

joostlek commented on tests/components/swiss_public_transport/test_config_flow.py:
```diff
@@ -0,0 +1,295 @@
+    assert result["handler"] == "swiss_public_transport"
+    assert result["data_schema"] == config_flow.DATA_SCHEMA
+
+    with patch(
+        "homeassistant.components.swiss_public_transport.config_flow.OpendataTransport",
+        autospec=True,
+    ) as mock_OpendataTransport:
+        mock_OpendataTransport_instance = mock_OpendataTransport.return_value
+        mock_OpendataTransport_instance.async_get_data.return_value = True
```
I think this can be shorter and a bit more readable. Can you maybe try this:

```suggestion
    with patch(
        "homeassistant.components.swiss_public_transport.config_flow.OpendataTransport.async_get_data",
    ):
```

This can require some tinkering, so feel free to hmu on Discord to have a chat about it

miaucl committed:
parametrize unit tests

gjohansson-ST commented on homeassistant/components/swiss_public_transport/strings.json:
```diff
@@ -0,0 +1,36 @@
+  "issues": {
+    "deprecated_yaml_import_issue_cannot_connect": {
+      "title": "The swiss public transport YAML configuration import cannot connect to server",
+      "description": "Configuring swiss public transport using YAML is being removed but there was an connection error importing your YAML configuration."
```
What should the user do?

gjohansson-ST commented on tests/components/swiss_public_transport/test_config_flow.py:
```diff
@@ -0,0 +1,220 @@
+        (IndexError(), "unknown"),
+    ],
+)
+async def test_flow_user_init_data_unknown_error_and_recover(
```
Seems same as the test above but also recovers which is good?

miaucl replied:
It was asked by @joostlek to show that the occurrence of any type of error does not hinder the completion of the config flow.

gjohansson-ST replied:
That's correct. I meant you could delete the test above as it's testing same thing as this one (which also recovers)

gjohansson-ST commented on homeassistant/components/swiss_public_transport/strings.json:
```diff
@@ -0,0 +1,36 @@
+{
+  "config": {
+    "error": {
+      "cannot_connect": "Cannot connect to server",
+      "bad_config": "Request failed due to bad config: Check at [stationboard](http://transport.opendata.ch/examples/stationboard.html) if your station names are valid",
+      "unknown": "An unknown error was raised by python-opendata-transport"
```
These should also be in `abort`

gjohansson-ST commented on homeassistant/components/swiss_public_transport/strings.json:
```diff
@@ -0,0 +1,36 @@
+      "unknown": "An unknown error was raised by python-opendata-transport"
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```
```suggestion
      "already_configured": "[%key:common::config_flow::abort::already_configured_service%]"
```

miaucl commented on homeassistant/components/swiss_public_transport/__init__.py:
```diff
@@ -1 +1,65 @@
+            f"Setup failed for entry '{start} {destination}' with invalid data"
+        ) from e
+
+    hass.data[DOMAIN][entry.entry_id] = opendata
```
It was already set further up, but not used.

miaucl commented on homeassistant/components/swiss_public_transport/strings.json:
```diff
@@ -0,0 +1,36 @@
+          "to": "End station"
+        },
+        "description": "Provide start and end station for your connection\n\nCheck here for valid stations: [stationboard](http://transport.opendata.ch/examples/stationboard.html)",
+        "title": "Setup connection"
```
Maybe just the name of the integration?

gjohansson-ST replied:
It will set the name if you omit this from strings

miaucl committed:
improve strings

miaucl commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -4,37 +4,40 @@
+    DOMAIN_HUMAN_READABLE,
+)
 
+_LOGGER = logging.getLogger(__name__)
```
Used on line 172

miaucl committed:
add missing aborts

miaucl commented:
It says for me, that there are still changes that need to be addressed, but I consider all the comments resolved. Is there something I am missing?

joostlek commented:
As long as you marked it as ready for review it's good! I'll have a look later today or tomorrow

joostlek commented:
Please also add the `__init__.py` to the .coveragerc

miaucl committed:
update coverage ignore

joostlek commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -0,0 +1,100 @@
+            except OpendataTransportError:
+                errors["base"] = "bad_config"
+            except Exception:  # pylint: disable=broad-except
+                errors["base"] = "unknown"
```
```suggestion
                _LOGGER.exception("Unknown error")
                errors["base"] = "unknown"
```

joostlek commented on tests/components/swiss_public_transport/test_config_flow.py:
```diff
@@ -0,0 +1,229 @@
+        (IndexError(), "unknown"),
+    ],
+)
+async def test_flow_user_init_data_cannot_connect_error(
+    hass: HomeAssistant, raise_error, text_error
+) -> None:
+    """Test cannot_connect errors."""
+    with patch(
+        "homeassistant.components.swiss_public_transport.config_flow.OpendataTransport.async_get_data",
+        autospec=True,
+        side_effect=raise_error,
+    ):
+        result = await hass.config_entries.flow.async_init(
+            config_flow.DOMAIN, context={"source": "user"}
+        )
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            user_input=MOCK_DATA_STEP,
+        )
+
+        assert result["type"] == "form"
+        assert result["errors"]["base"] == text_error
+
+
+@pytest.mark.parametrize(
+    ("raise_error", "text_error"),
+    [
+        (OpendataTransportConnectionError(), "cannot_connect"),
+        (OpendataTransportError(), "bad_config"),
+        (IndexError(), "unknown"),
+    ],
+)
+async def test_flow_user_init_data_unknown_error_and_recover(
```
Aren't these the same tests?

miaucl committed:
remove redundant test

gjohansson-ST commented on homeassistant/components/swiss_public_transport/__init__.py:
```diff
@@ -1 +1,64 @@
+async def async_setup_entry(
+    hass: core.HomeAssistant, entry: config_entries.ConfigEntry
+) -> bool:
+    """Set up platform from a ConfigEntry."""
```
```suggestion
    """Set up Swiss public transport from a config entry."""
```

gjohansson-ST commented on homeassistant/components/swiss_public_transport/config_flow.py:
```diff
@@ -0,0 +1,101 @@
+"""Config flow for transport.opendata.ch."""
```
```suggestion
"""Config flow for swiss_public_transport."""
```

gjohansson-ST commented on homeassistant/components/swiss_public_transport/const.py:
```diff
@@ -0,0 +1,19 @@
+
+DOMAIN = "swiss_public_transport"
+
+ATTR_DEPARTURE_TIME1 = "next_departure"
+ATTR_DEPARTURE_TIME2 = "next_on_departure"
+ATTR_DURATION = "duration"
+ATTR_PLATFORM = "platform"
+ATTR_REMAINING_TIME = "remaining_time"
+ATTR_START = "start"
+ATTR_TARGET = "destination"
+ATTR_TRAIN_NUMBER = "train_number"
+ATTR_TRANSFERS = "transfers"
+ATTR_DELAY = "delay"
```
These are only used in `sensor.py` so should remain there and not be moved to here.

gjohansson-ST commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -47,31 +49,65 @@
+
+    start = config_entry.data[CONF_START]
+    destination = config_entry.data[CONF_DESTINATION]
+    name = config_entry.data.get(CONF_NAME, f"{start} {destination}")
```
```suggestion
    name = config_entry.title
```

gjohansson-ST commented on homeassistant/components/swiss_public_transport/sensor.py:
```diff
@@ -47,31 +49,65 @@
-    async_add_entities([SwissPublicTransportSensor(opendata, start, destination, name)])
 
 
 class SwissPublicTransportSensor(SensorEntity):
```
Please add unique id and a device (service) in a follow-up PR

miaucl replied:
I am already working at an extension of this integration and will consider it there. This PR is only to add config_flow functionality.

gjohansson-ST reviewed (changes requested):
Some minor last bits then it should be good to go

miaucl committed:
minor clean up of constants

gjohansson-ST reviewed (approved):
Thanks @miaucl 👍

MartinHjelmare commented on homeassistant/components/swiss_public_transport/__init__.py:
```diff
@@ -1 +1,64 @@
+            f"Timeout while connecting for entry '{start} {destination}'"
+        ) from e
+    except OpendataTransportError as e:
+        _LOGGER.error(
```
We shouldn't both log an error and raise an exception. Just raise the exception. The exception argument will be logged by default.

miaucl replied:
@MartinHjelmare https://github.com/home-assistant/core/pull/111741

MartinHjelmare commented on homeassistant/components/swiss_public_transport/__init__.py:
```diff
@@ -1 +1,64 @@
+        ) from e
+    except OpendataTransportError as e:
+        _LOGGER.error(
+            "Setup failed for entry '%s %s', check at http://transport.opendata.ch/examples/stationboard.html if your station names are valid",
```
Please break long strings around max 88 characters per line.
